### ISSUE TITLE
Upgrade Bitcoin app version if above minimum but below Taproot minimum

### ISSUE THREAD

**Issue by darosior** [wizardsardine/liana#1135]
Enough people have been testing https://github.com/darosior/ledger_installer that i feel comfortable starting to integrate pieces into Liana. A good small first step would be to make it possible for Ledger users with a Bitcoin app whose version is below the version which introduced Taproot support to upgrade to the latest Bitcoin app. Note however those users must already have a Liana-compatible app (above our minimum supported version) otherwise the update of the app would also necessitates a firmware upgrade, which is not implemented at the moment (and not the best candidate for a first step).

**Comment by pythcoiner** [comment]
how do you see this? call ledger_installer gui as an external process?

**Comment by darosior** [comment]
No, i think the GUI should just have a button letting the user upgrade in case they choose to use Taproot but their device's Bitcoin app is too old. This would just call the library's function directly.

**Comment by pythcoiner** [comment]
open a modal to show hint (accept on device) and progress (bubling sardine?)?

**Comment by darosior** [comment]
I'm not sure yet how it should be integrated, but i think it should be minimal. For instance if we choose to have the upgrade button inside the signing devices list item, then opening a pop-up would be a pop-up on top of a pop-up which might not be ideal.

Maybe when you start tackling this, start by sketching out the idea and propose it here before moving on with the implementation?

**Comment by pythcoiner** [comment]
I've started look at this, i think we should at least have the feature in 3 location:
 - In installer "Select a signing device" modal:
  
![Image](https://github.com/user-attachments/assets/bec1db57-ef06-46de-9605-6f5802114d21)

- In the installer "Register descriptor step" (3/6)

![Image](https://github.com/user-attachments/assets/537bd6bb-1664-4a54-b528-646109e68007)

- In wallet "Select signing device to sign with":

![Image](https://github.com/user-attachments/assets/6da7b42d-058e-4abc-b948-fbd40b58297b)

The way i think to achieve this is to grey-out (disable) the signing device selection button and add inside it add an " Upgrade my ledger" button that will trigger the upgrade process while keeping the modal opened.

Points to consider
 - i think we should publish [ledger_manager](https://github.com/darosior/ledger_installer/blob/master/ledger_manager/Cargo.toml) on crates.io
 - i think we should not let user leave the modal if an upgrade is in progress.
 - i think we should display a log of Hints & infos about the upgrade process.
 - Should we allow several signing devices to upgrade at same time?

**Comment by darosior** [comment]
I was wondering why it would be necessary after the installer step, but it may be the case that a participant who shared his xpub may not have the appropriate version. In this case we should indeed have a version check before registering the descriptor and instruct the user to upgrade if the version is too old. But why having it at signing time? You may only ever sign with a device upon which was registered the descriptor, so having the check there seem sufficient?

**Comment by darosior** [comment]
> i think we should publish [ledger_manager](https://github.com/darosior/ledger_installer/blob/master/ledger_manager/Cargo.toml) on crates.io

Sure, we can do it but it's not a blocker. Especially if it turns out we need to adjust a couple things.

> i think we should not let user leave the modal if an upgrade is in progress.

Probably, yes. We can never ensure the user won't interrupt the upgrade though.

> i think we should display a log of Hints & infos about the upgrade process.

Ideally we'd also show the progress on the UI but i can understand how it can be difficult.

> Should we allow several signing devices to upgrade at same time?

Why not, but let's keep it simple for now. (Also this conflicts with your second statement of being stuck on the modal.)

**Comment by darosior** [comment]
> The way i think to achieve this is to grey-out (disable) the signing device selection button and add inside it an " Upgrade my ledger" button that will trigger the upgrade process while keeping the modal opened:

It's not clear for the user when they should click on the button nor is it technically correct because of the size restriction.

Instead what do you think of:
1. Greying out the list item
2. Displaying an attached yellow banner just below it (in our usual warning yellow) with mention "Your Ledger's Bitcoin app is too old to be used in Taproot descriptor." aligned on the left, and then an "upgrade" button aligned on the right.

I've tried to draft what i have in mind to be clearer:
![signal-2024-07-23-094130](https://github.com/user-attachments/assets/dd2c454b-9765-48d0-9b55-9831cb3e5929)

**Comment by pythcoiner** [comment]
> I was wondering why it would be necessary after the installer step, but it may be the case that a participant who shared his xpub may not have the appropriate version. In this case we should indeed have a version check before registering the descriptor and instruct the user to upgrade if the version is too old. But why having it at signing time? You may only ever sign with a device upon which was registered the descriptor, so having the check there seem sufficient?

In the case an user have broken his ledger, replace by a new one, import his private key into, as Liana store the PoR, i think it will be detected as registered. So in this case user can end up to signing step w/o Liana told him the device is not registered.

**Comment by darosior** [comment]
> user can only interrupt on hardware side and i think we can detect it and then allow user to leave the modal

They can just close the software altogether. Preventing to close the modal is just to prevent a simple mistake imo (and might be worth it), but it's not possible to protect the user from themselves in this case.

**Comment by pythcoiner** [comment]
> I've tried to draft what i have in mind to be clearer: ![signal-2024-07-23-094130](https://private-user-images.githubusercontent.com/22457751/351233198-dd2c454b-9765-48d0-9b55-9831cb3e5929.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3MjE5NDYsIm5iZiI6MTcyMTcyMTY0NiwicGF0aCI6Ii8yMjQ1Nzc1MS8zNTEyMzMxOTgtZGQyYzQ1NGItOTc2NS00OGQwLTliNTUtOTgzMWNiM2U1OTI5LmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyM1QwODAwNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NzllMTE3MGUyMTlkM2E4MDkwNzk4OGRhNzkxMTA4YWQ4ZjE1YzVmZDMxODVmNjI2YjVmMThlYmQ2YjhhY2NjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.t_Zde29Iz4K4mrlfOy5KLMmALBHn36Y5mpMC-F4gjSQ)

it's not exactly what you have drafted, but what do you think about this (modulo the color theme that should been adjusted for the button at least):

![Image](https://github.com/user-attachments/assets/ed408a0e-c492-4b8e-9476-7caf5caae4e1)

(i do not see an easy way to make the yellow banner 'attached' to the device button, so I put it inside the button, as if below,the user can be confused if there is 2 following ledger in the list)

**Comment by pythcoiner** [comment]
or maybe something like this?


![Image](https://github.com/user-attachments/assets/71845bf9-0e13-4fcc-9a52-d9ea24786aad)

**Comment by pythcoiner** [comment]
or like this:

![Image](https://github.com/user-attachments/assets/df4950b0-fe3f-4d96-aa73-43098f6fb022)

**Comment by darosior** [comment]
The last one looks good to me (besides using a white font color for the Upgrade button).

-------- Original Message --------

**Comment by pythcoiner** [comment]
i think we should also hae the upgrade feature here, for the same reasons we need at signing step.

![image](https://github.com/user-attachments/assets/6925a392-3d1e-4c99-8844-37878d8ce294)
